R/eval.R

```diff
@@ -37,27 +52,76 @@ rust_eval <- function(code, env = parent.frame(), ...) {
     ui_throw("decoy function; should never be called.")
   }
 
+  # Snippet hash is constructed from the Rust source code and
+  # a unique identifier of the compiled dll.
+  # Every time any rust code is dynamically compiled,
+  # `the$count` is incremented.
+  # This ensures that any two (even bytewise-identical)
+  # Rust source code strings will have different
+  # hashes.
+  snippet_hash <- rlang::hash(list(the$count, code))
+
+  # The unique hash is then used to generate unique function names
+  fn_name <- glue("rextendr_rust_eval_fun_{snippet_hash}")
+
   # wrap code into Rust function
   code_wrapped <- glue(r"(
-fn rextendr_rust_eval_fun() -> Result<Robj> {{
+fn {fn_name}() -> Result<Robj> {{
   let x = {{
     {code}
   }};
   Ok(x.into())
 }}
 )")
-  out <- rust_function(code = code_wrapped, ...)
-  result <- do.call(rextendr_rust_eval_fun, list(), envir = env)
-  dyn.unload(out[["path"]])
-
-  # Attempt to figure out whether the Rust code returns a result or not,
-  # and make the result invisible or not accordingly. This regex approach
-  # is not perfect, but since it only affects the visibility of the result
-  # that's Ok. Worst case scenario a result that should be invisible is
-  # shown as visible.
-  if (grepl(".*;\\s*$", code, perl = TRUE)) {
-    invisible(result)
+
+  has_no_return <- grepl(".*;\\s*$", code, perl = TRUE)
```

Comment:
Looks like the comment explaining the grepl line got lost in the shuffle. Could you restore it?

Reply:
Yeah, sure, I indeed lost it.